editor/src/clj/editor/editor_extensions/coerce.clj

```diff
@@ -0,0 +1,322 @@
+
+  Optional kv-args:
+    :min-count    minimum number of items in the collection
+    :distinct     if true, ensures that the result does not have repeated items"
+  [item-coercer & {:keys [min-count distinct]}]
+  (let [f (fn coerce-coll-of [vm ^LuaValue x]
+            (if (.istable x)
+              (let [acc
+                    (reduce
+                      (fn [acc ^Varargs varargs]
+                        (let [lua-k (.arg1 varargs)]
+                          (if (.isinttype lua-k)
+                            (let [i (dec (.tolong lua-k))]  ;; 1-indexed to 0-indexed
+                              (if (neg? i)
+                                acc                         ;; non-array index, skip
+                                (let [value-or-failure (item-coercer vm (.arg varargs 2))]
+                                  (if (failure? value-or-failure)
+                                    (reduced value-or-failure)
+                                    (let [acc (loop [acc acc]
+                                                (if (< (count acc) i)
+                                                  (recur (conj! acc ::not-found))
+                                                  acc))]
+                                      (assoc! acc i value-or-failure))))))
+                            acc)))
+                      (transient [])
+                      (vm/lua-table-reducer vm x))]
+                (if (failure? acc)
+                  acc
+                  ;; check holes
+                  (let [n (count acc)
+                        first-hole-index (loop [i 0]
+                                           (cond
+                                             (= i n) nil
+                                             (identical? ::not-found (acc i)) i
+                                             :else (recur (inc i))))]
+                    (if first-hole-index
+                      (let [first-hole-index (long first-hole-index)
+                            coerced-nil (item-coercer vm LuaValue/NIL)]
+                        (if (failure? coerced-nil)
+                          coerced-nil
+                          (loop [acc (assoc! acc first-hole-index coerced-nil)
+                                 i (inc first-hole-index)]
+                            (cond
+                              (= i n) (persistent! acc)
+                              (identical? ::not-found (acc i)) (recur (assoc! acc i coerced-nil) (inc i))
+                              :else (recur acc (inc i))))))
+                      (persistent! acc)))))
+              (failure x "is not an array")))]
+    (cond-> f
+
+            min-count
+            (wrap-with-pred
+              #(<= ^long min-count (count %))
+              (str "needs at least " min-count " element" (when (< 1 ^long min-count) "s")))
+
+            distinct
+            (wrap-with-pred #(apply distinct? %) "should not have repeated elements"))))
```

Comment:
This looks like a fairly involved process in order to support coercing a sparse hashmap into a non-sparse vector. I'm not sure is it worth the effort?

I think we will have some APIs that expect vectors and some that expect hashmaps. You could argue a sparse Lua table should not be able to be coerced to a vector at all, and that we should only coerce the array part of Lua tables and ignore the hashmap part when we need a vector.

I worry being flexible will significantly slow down marshaling for very little real benefit. Better to not allow it at all.

Reply:
Yes, I'll simplify the code to use the Lua array semantics.